Consteval method in C++20/Clang unhappy?

@benthevining I think we went a bit overboard when trying to solve the perfetto C++17 deprecation, maybe you could take a look?

I'm having trouble compiling my parent project in C++20, which has the following set:

```
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

It's only a problem in Release, not in Debug.

I thought maybe the problem was `CMAKE_CXX_STANDARD` in the main perfetto CMakeLists.txt and I should swap to 

```
# set a *minimum* of C++17, but allow higher
target_compile_features(perfetto PRIVATE cxx_std_17)
```

But that hasn't resolved the issue.

I get the following error (Release only):

```
clang: error: unable to execute command: Segmentation fault: 11
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: x86_64-apple-darwin22.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
clang: note: diagnostic msg: Error generating preprocessed source(s) - cannot generate preprocessed source with multiple -arch options.
[82/144] Building CXX object CMakeFiles/SineMachine_v11.dir/JUCE/modules/juce_core/juce_core.mm.o
```

This was also reproduced by someone on the forum.

Maybe the problem is the fact that `perfetto::StaticString` isn't consteval friendly, or shouldn't be used.

When I tried to remove that I got a perfetto assert:

`/Users/sudara/projects/sinemachine/cmake-build-debug/_deps/perfetto-src/sdk/perfetto.h:17421:3: error: static_assert failed due to requirement 'IsValidEventNameType<char *, void>::value' "Event names must be static strings. To use dynamic event names, see https://perfetto.dev/docs/instrumentation/track-events#dynamic-event-names"`

The normal supported perfetto route is to pass a string literal like `TRACE_EVENT ("component", "myName")`

--- a/melatonin_perfetto/melatonin_perfetto.h
+++ b/melatonin_perfetto/melatonin_perfetto.h
@@ -16,7 +16,7 @@ END_JUCE_MODULE_DECLARATION
 #pragma once
 
 #ifndef PERFETTO
-    #define PERFETTO 1
+    #define PERFETTO 0
 #endif
 
 #if PERFETTO
@@ -172,7 +172,7 @@ namespace melatonin
 
                 // MSVC has an additional identifier after the return type: __cdecl
                 if (src[i + 1] == '_')
-                    i += 8; // skip the space and __cdecl
+                    i += 8; // skip __cdecl and the space afterwards
 
                 size_t j = 0;
 

--- a/melatonin_perfetto/melatonin_perfetto.h
+++ b/melatonin_perfetto/melatonin_perfetto.h
@@ -3,7 +3,7 @@ BEGIN_JUCE_MODULE_DECLARATION
 
  ID:                 melatonin_perfetto
  vendor:             Sudara
- version:            1.0.0
+ version:            1.1.0
  name:               Melatonin Perfetto
  description:        Perfetto module for JUCE
  license:            MIT

--- a/melatonin_perfetto/melatonin_perfetto.h
+++ b/melatonin_perfetto/melatonin_perfetto.h
@@ -110,67 +110,125 @@ private:
             output->setPosition (0);
             output->write (&trace_data[0], trace_data.size() * sizeof (char));
 
-            DBG("Wrote perfetto trace to: " + childFile.getFullPathName());
+            DBG ("Wrote perfetto trace to: " + childFile.getFullPathName());
 
             return childFile;
         }
 
-        DBG("Failed to write perfetto trace file. Check for missing permissions.");
+        DBG ("Failed to write perfetto trace file. Check for missing permissions.");
         jassertfalse;
-        return juce::File{};
+        return juce::File {};
     }
 
     std::unique_ptr<perfetto::TracingSession> session;
 };
 
-    /* https://stackoverflow.com/a/72627251
-       Accepts an array of characters and builds a new truncated array
+/*
 
-       Without this, deriving the function name produces an ugly string such as
-            auto AudioProcessor::processBlock(juce::AudioBuffer<float> &, juce::MidiBuffer &)::(anonymous class)::operator()()::(anonymous class)::operator()(uint32_t) const
+   There be dragons here. Serious C++ constexpr dragons.
 
-       With this, we get a nice string such as
-            AudioProcessor::processBlock
-    */
+   Deriving the function name produces an ugly string:
+        auto AudioProcessor::processBlock(juce::AudioBuffer<float> &, juce::MidiBuffer &)::(anonymous class)::operator()()::(anonymous class)::operator()(uint32_t) const
 
-#undef MELATONIN_CONSTEVAL
+   Without dragons, trying to trim the string results in it happening at runtime.
 
-// Guarantee consteval where possible
-#if __cplusplus >= 202002L
-    #define MELATONIN_CONSTEVAL consteval
-#else
-    #define MELATONIN_CONSTEVAL constexpr
-#endif
+   With dragons, we get a nice compile time string:
+        AudioProcessor::processBlock
 
-template <auto N>
-MELATONIN_CONSTEVAL auto class_and_method_name (char const (&src)[N]) noexcept
+*/
+
+namespace melatonin
 {
-    std::array<char, N> res = {};
-    for (size_t i = 0; i < N; ++i)
-    {
-        // wait until after the return type (first space in the string)
-        if (src[i] == ' ')
-        {
-            ++i; // skip the space
-            size_t j = 0;
+    // This wraps our compile time strings (coming from PRETTY_FUNCTION, etc) in lambdas
+    // so that they can be used as template parameters
+    // This lets us trim the strings while still using them at compile time
+    // https://accu.org/journals/overload/30/172/wu/#_idTextAnchor004
+    #define WRAP_COMPILE_TIME_STRING(x) [] { return (x); }
+    #define UNWRAP_COMPILE_TIME_STRING(x) (x)()
 
-            // build result, stop when we hit the arguments
-            while (src[i] != '(' && i < N && j < N)
+    template <typename CompileTimeLambdaWrappedString>
+    constexpr auto compileTimePrettierFunction (CompileTimeLambdaWrappedString wrappedSrc)
+    {
+    // if we're C++20 or higher, ensure we're compile-time
+    #if __cplusplus >= 202002L
+        // This should never assert, but if so,  report it on this issue:
+        // https://github.com/sudara/melatonin_perfetto/issues/13#issue-1558171132
+        if (!std::is_constant_evaluated())
+            jassertfalse;
+    #endif
+
+        constexpr auto src = UNWRAP_COMPILE_TIME_STRING (wrappedSrc);
+        constexpr auto size = std::string_view (src).size(); // -1 to ignore the /0
+        std::array<char, size> result {};
+
+        // loop through the source, building a new truncated array
+        // see: https://stackoverflow.com/a/72627251
+        for (size_t i = 0; i < size; ++i)
+        {
+            // wait until after the return type (first space in the string)
+            if (src[i] == ' ')
             {
-                res[j] = src[i];
-                ++i; // increment character in source
-                ++j; // increment character in result
+                ++i; // skip the space
+
+                // MSVC has an additional identifier after the return type: __cdecl
+                if (src[i + 1] == '_')
+                    i += 8; // skip the space and __cdecl
+
+                size_t j = 0;
+
+                // build result, stop when we hit the arguments
+                while (src[i] != '(' && i < size && j < size)
+                {
+                    result[j] = src[i];
+                    ++i; // increment character in source
+                    ++j; // increment character in result
+                }
+                return result;
             }
-            return res;
         }
+        return result;
     }
-    return res;
+
+    namespace test
+    {
+        // a lil test helper so we don't go crazy
+        template <size_t sizeResult, size_t sizeTest>
+        constexpr bool strings_equal (const std::array<char, sizeResult>& result, const char (&test)[sizeTest])
+        {
+            // sanity check
+            static_assert (sizeTest > 1);
+            static_assert (sizeResult + 1 >= sizeTest);
+
+            std::string_view resultView (result.data(), sizeTest);
+            std::string_view testView (test, sizeTest);
+            return testView == resultView;
+        }
+
+        // testing at compile time isn't fun (no debugging) so dumb things like this help:
+        constexpr std::array<char, 10> main { "main" };
+        static_assert (strings_equal (main, "main"));
+
+        // ensure the return type is removed
+        static_assert (strings_equal (compileTimePrettierFunction (WRAP_COMPILE_TIME_STRING ("int main")), "main"));
+
+        // clang example
+        static_assert (strings_equal (compileTimePrettierFunction (WRAP_COMPILE_TIME_STRING ("void AudioProcessor::processBlock(juce::AudioBuffer<float> &, juce::MidiBuffer &)::(anonymous class)::operator()()::(anonymous class)::operator()(uint32_t) const")), "AudioProcessor::processBlock"));
+
+        // msvc example
+        static_assert (strings_equal (compileTimePrettierFunction (WRAP_COMPILE_TIME_STRING ("void __cdecl AudioProcessor::processBlock(class juce::AudioBuffer<float,0,0> &,class juce::MidiBuffer &)")), "AudioProcessor::processBlock"));
+    }
 }
 
-#undef MELATONIN_CONSTEVAL
-
-    #define TRACE_DSP(...) TRACE_EVENT ("dsp", perfetto::StaticString (class_and_method_name (PERFETTO_DEBUG_FUNCTION_IDENTIFIER()).data()), ##__VA_ARGS__)
-    #define TRACE_COMPONENT(...) TRACE_EVENT ("component", perfetto::StaticString (class_and_method_name (PERFETTO_DEBUG_FUNCTION_IDENTIFIER()).data()), ##__VA_ARGS__)
+    // Et voilà! Our nicer macros.
+    // This took > 20 hours, hope the DX is worth it...
+    // The separate constexpr calls are required for `compileTimePrettierFunction` to remain constexpr
+    // in other words, they can't be inline with perfetto::StaticString, otherwise it will go runtime
+    #define TRACE_DSP(...)                                                                                                            \
+        constexpr auto pf = melatonin::compileTimePrettierFunction (WRAP_COMPILE_TIME_STRING (PERFETTO_DEBUG_FUNCTION_IDENTIFIER())); \
+        TRACE_EVENT ("dsp", perfetto::StaticString (pf.data()), ##__VA_ARGS__)
+    #define TRACE_COMPONENT(...)                                                                                                      \
+        constexpr auto pf = melatonin::compileTimePrettierFunction (WRAP_COMPILE_TIME_STRING (PERFETTO_DEBUG_FUNCTION_IDENTIFIER())); \
+        TRACE_EVENT ("component", perfetto::StaticString (pf.data()), ##__VA_ARGS__)
 
 #else // if PERFETTO
     #define TRACE_EVENT_BEGIN(category, ...)

--- a/melatonin_perfetto/melatonin_perfetto.h
+++ b/melatonin_perfetto/melatonin_perfetto.h
@@ -16,7 +16,7 @@ END_JUCE_MODULE_DECLARATION
 #pragma once
 
 #ifndef PERFETTO
-    #define PERFETTO 0
+    #define PERFETTO 1
 #endif
 
 #if PERFETTO
@@ -177,11 +177,19 @@ namespace melatonin
                 size_t j = 0;
 
                 // build result, stop when we hit the arguments
-                while (src[i] != '(' && i < size && j < size)
+                // clang and gcc use (, MSVC uses <
+                while ((src[i] != '(' && src[i] != '<') && i < size && j < size)
                 {
                     result[j] = src[i];
                     ++i; // increment character in source
                     ++j; // increment character in result
+                }
+
+                // ugly clean up after msvc, we want to remove the extra :: before <lambda_1>
+                if (src[i] == '<')
+                {
+                    result[j] = ' ';
+                    result[j-1] = ' ';
                 }
                 return result;
             }
@@ -215,7 +223,7 @@ namespace melatonin
         static_assert (strings_equal (compileTimePrettierFunction (WRAP_COMPILE_TIME_STRING ("void AudioProcessor::processBlock(juce::AudioBuffer<float> &, juce::MidiBuffer &)::(anonymous class)::operator()()::(anonymous class)::operator()(uint32_t) const")), "AudioProcessor::processBlock"));
 
         // msvc example
-        static_assert (strings_equal (compileTimePrettierFunction (WRAP_COMPILE_TIME_STRING ("void __cdecl AudioProcessor::processBlock(class juce::AudioBuffer<float,0,0> &,class juce::MidiBuffer &)")), "AudioProcessor::processBlock"));
+        static_assert (strings_equal (compileTimePrettierFunction (WRAP_COMPILE_TIME_STRING ("void __cdecl AudioProcessor::processBlock::<lambda_1>::operator")), "AudioProcessor::processBlock"));
     }
 }
 

--- a/melatonin_perfetto/melatonin_perfetto.h
+++ b/melatonin_perfetto/melatonin_perfetto.h
@@ -185,11 +185,10 @@ namespace melatonin
                     ++j; // increment character in result
                 }
 
-                // ugly clean up after msvc, we want to remove the extra :: before <lambda_1>
+                // really ugly clean up after msvc, remove the extra :: before <lambda_1>
                 if (src[i] == '<')
                 {
-                    result[j] = ' ';
-                    result[j-1] = ' ';
+                    result[j-2] = '\0';
                 }
                 return result;
             }
@@ -205,7 +204,7 @@ namespace melatonin
         {
             // sanity check
             static_assert (sizeTest > 1);
-            static_assert (sizeResult + 1 >= sizeTest);
+            static_assert (sizeResult + 1 >= sizeTest); // +1 for the /0
 
             std::string_view resultView (result.data(), sizeTest);
             std::string_view testView (test, sizeTest);

